File: LeetCode/76-minimum-window-substring/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {

  public static String minWindow(String s, String t) {
    // Edge Case: Return empty string if either argument is empty
    if (s.length() == 0 || t.length() == 0) {
      return "";
    }

    // Frequency map for characters in T
    Map<Character, Integer> targetFreq = new HashMap<>();
    for (char c : t.toCharArray()) {
      targetFreq.put(c, targetFreq.getOrDefault(c, 0) + 1);
    }

    // Two pointers for the sliding window
    int left = 0, right = 0;
    int required = targetFreq.size();
    int formed = 0;

    // Frequency map for the current window
    Map<Character, Integer> windowCounts = new HashMap<>();
    // Variables to keep track of the smallest window
    int[] result = { -1, 0, 0 }; // length, left, right

    while (right < s.length()) {
      // Add character from the right to the window
      char c = s.charAt(right);
      windowCounts.put(c, windowCounts.getOrDefault(c, 0) + 1);

      // Check if current window has enough occurrences of the current character
      if (targetFreq.containsKey(c) && windowCounts.get(c).intValue() == targetFreq.get(c).intValue()) {
        formed++;
      }

      // Contract the window until it is no longer valid
      while (left <= right && formed == required) {
        c = s.charAt(left);
        // Save the smallest window
        if (result[0] == -1 || right - left + 1 < result[0]) {
          result[0] = right - left + 1;
          result[1] = left;
          result[2] = right;
        }

        // Remove character from the left from the window
        windowCounts.put(c, windowCounts.get(c) - 1);
        if (targetFreq.containsKey(c) && windowCounts.get(c).intValue() < targetFreq.get(c).intValue()) {
          formed--;
        }
        left++;
      }
      right++;
    }

    return result[0] == -1 ? "" : s.substring(result[1], result[2] + 1);
  }
}

File: LeetCode/76-minimum-window-substring/solution.ts
function minWindow(s: string, t: string): string {
  if (s.length === 0 || t.length === 0) {
    return "";
  }

  // Frequency map for characters in T
  const targetFreq: Map<string, number> = new Map();
  for (const char of t) {
    targetFreq.set(char, (targetFreq.get(char) || 0) + 1);
  }

  let left = 0,
    right = 0;
  let required = targetFreq.size;
  let formed = 0;

  // Frequency map for the current window
  const windowCounts: Map<string, number> = new Map();
  let result: [number, number, number] = [-1, 0, 0]; // [length, left, right]

  while (right < s.length) {
    const char = s[right];
    windowCounts.set(char, (windowCounts.get(char) || 0) + 1);

    if (
      targetFreq.has(char) &&
      windowCounts.get(char) === targetFreq.get(char)
    ) {
      formed++;
    }

    while (left <= right && formed === required) {
      const startChar = s[left];

      // Save the smallest window
      if (result[0] === -1 || right - left + 1 < result[0]) {
        result = [right - left + 1, left, right];
      }

      windowCounts.set(startChar, windowCounts.get(startChar)! - 1);
      if (
        targetFreq.has(startChar) &&
        windowCounts.get(startChar)! < targetFreq.get(startChar)!
      ) {
        formed--;
      }

      left++;
    }

    right++;
  }

  return result[0] === -1 ? "" : s.substring(result[1], result[2] + 1);
}
